### DISCUSSION TITLE
Color groups

### DISCUSSION THREAD

**Discussion by T14D3** [YimMenu/YimMenu#1170] [Ideas]
Hello,
since my previous implementation of adding customizable GUI colors had some negative feedback regarding the "groups" I chose to put different colors in, I'd like to get some input on what sort of areas you thing should be included as the "same" color.
For me the obvious ones would be:
- Background
- Text
- Buttons default
- Buttons Hover and Active (same color, different brightness)
- Checkbox tick
- Checkmark background
- Scrollbar default
- Scrollbar Hover and Active (same color, different brightness)

If you have any ideas for this, please let me know!

**Comment by Seanghost117** [comment]
I like your direction with this one, it's a lot more specific and beneficial to the user. Although right now I'm working on a (temporary) ui PR that should change the whole layout, make the window a more constant size, change positions and also may override the default checkboxes, will include options for background images and support for the tabs both horizontal and vertical that I'm attempting to implement at the moment. I'll either make a new discussion thread or a PR will open for this soon. I say temporary because it will all be moved to CEF at some point in the future

**Reply by tupoy-ya** [reply]
Out of all problems with YimMenu the main problem remains - it's not chrome.
Everything should be chrome, it doesn't mater what, a game cheat, text editor, browser, game library, messenger, it all has to be chrome.
That's the main problem of humanity - we don't have enough chrome!

EDIT: You can't hide comments as off-topic for some reason.

**Reply by xiaoxiao921** [reply]
That's your whole argument? Maybe because it's everywhere means because it's actually good don't you think? If your argument was because of the size bloat which I totally understand (libcef is basically 100mb+ big) here it's not even an issue due to it being already shipped by the social club in game overlay

**Reply by tupoy-ya** [reply]
It's just personal preference i prefer to use ImGui over a web browser when making a game UI.

Also while SC overlay does ship libcef i sometimes use Goldberg emulator to play singleplayer without Social Club, and right now making YimMenu work without SocialClub is as simple as removing a few hooks but after adding libcef to the menu you will have to rewrite the entire UI to make it work or use some custom libcef.

And what are the benefits of having libcef in the menu?

**Reply by xiaoxiao921** [reply]
Much better UI, easier and faster to develop complex layout if needed (everyone know html and css)

**Reply by GeopJr** [reply]
As an outsider to the GTA stuff, cef's benefits would be:
- Easier testing; we could test the UI separately
- Easier development; we could design & work on the UI without actually having GTA or building yimmenu
- Organization (I assume); right now imgui is sprinkled all over the source
- Keyboard navigation, accessibility, sizing, rtl handling
- room for innovation (?); (I assume) we could update the ui dynamically (like the translations), we could play audio, video etc, we could add a built in google fonts picker

**Reply by tupoy-ya** [reply]
As Wine user cef's benefits would be:
* Doesn't work
I have yet to see a single program with cef in it to work properly under Wine, this doesn't mean it's not possible but still.
Examples:
GTA5 Social Club overlay: Doesn't have a mouse cursor.
FiveM: Integrated VSCode freezes immediately.
Garry's Mod: Cef just doesn't start.
Steam: UI flickers a lot, didn't actually test just saw in a video.

As for what GeopJr have listed
* Easier testing; Yes
* Easier development; Yes
* Organization;  The best fix to unorganized code is to organize code, not switch to chromium
* Keyboard navigation: Can be done in ImGui or anything else
* Accessibility: Not sure what you mean by that
* Sizing: Can be done in ImGui or anything else
* rtl handling: Yes
* We could update the ui dynamically: We already can
* We could play audio: We don't need cef to do that
* We could play video: Can be done in ImGui although it's way easier with CEF, what is there to play in a GTA Menu though?
* We could add a built in google fonts picker: Yes

> Everyone know html and css

I don't, it's easier for me to make something in ImGui then to write HTML CSS

**Reply by Yimura** [reply]
Your UI already differs from YimMenu UI, I don't see how this would be a problem to you?

**Reply by GeopJr** [reply]
While I don't disagree with your point and believe we should do testing even on wine before actually switching, I'll bring some counter arguments:

While of course organization can be improved, cef would force everyone to be more organized with e.g. a global config for options. Even if the current UI gets extremely organized, a huge PR could easily get past it.

Keyboard navigation, sizing, media viewing etc. can be done in imgui but it's very easy to miss details like unusual display sizes, input devices or formats. On the same note, the inherited accessibility from the browser would make this easier, for example there's many ways to change the value of a slider: arrow keys, mouse wheel, num keys, up and down buttons etc

On dynamic updates: I meant something like having an html file that is missing a certain option and then pushing an update to it that adds it and all yimmenu instances get it without downloading a new version of yimmenu

**Reply by tupoy-ya** [reply]
> Your UI already differs from YimMenu UI, I don't see how this would be a problem to you?

It would be a pain to merge. Right now merging a new UI tab is as simple as just moving it from gui_service to an ImGui tab, but with cef i will have to remake the entire tab in ImGui.

> While of course organization can be improved, cef would force everyone to be more organized with e.g. a global config for options. Even if the current UI gets extremely organized, a huge PR could easily get past it.

A global config for options is `globals.hpp` maybe you meant something else and i'm just stupid.
A huge PR can easily break organization no mater what is used to draw the UI.

> Keyboard navigation, sizing, media viewing etc. can be done in imgui but it's very easy to miss details like unusual display sizes, input devices or formats. On the same note, the inherited accessibility from the browser would make this easier, for example there's many ways to change the value of a slider: arrow keys, mouse wheel, num keys, up and down buttons etc

Your UI already hard codes window sizes, I don't see how this would be a problem to you?
Well actually yes using browser elements will make keyboard navigation easier than default ImGui keyboard navigation, but both of them would require you to stand still and block all ingame movement while navigating the menu.

> On dynamic updates: I meant something like having an html file that is missing a certain option and then pushing an update to it that adds it and all yimmenu instances get it without downloading a new version of yimmenu

That would require a versioning system so that the old versions won't get buttons for functions they don't have.
And this will also require constant internet connection.


We should probably make a poll, where we outline all pros and cons to see what other people think.
